apps/io-wallet-user-func/src/android-key-attestation.ts

```diff
@@ -0,0 +1,127 @@
+import * as appInsights from "applicationinsights";
+import { X509Certificate } from "crypto";
+import * as E from "fp-ts/lib/Either";
+import * as RTE from "fp-ts/lib/ReaderTaskEither";
+import * as TE from "fp-ts/lib/TaskEither";
+import { pipe } from "fp-ts/lib/function";
+import { WalletInstanceValid } from "io-wallet-common/wallet-instance";
+
+import { AttestationServiceConfiguration } from "./app/config";
+import {
+  validateIssuance,
+  validateRevocation,
+} from "./infra/attestation-service/android/attestation";
+import {
+  WalletInstanceRepository,
+  getAllValidWalletInstances,
+  revokeUserWalletInstances,
+} from "./wallet-instance";
+
+type WalletInstanceValidWithAndroidCertificatesChain = {
+  deviceDetails: { platform: string; x509Chain: string[] };
+} & WalletInstanceValid;
+
+const hasAndroidCertificatesChain = (
+  walletInstance: WalletInstanceValid,
+): walletInstance is WalletInstanceValidWithAndroidCertificatesChain =>
+  !!walletInstance.deviceDetails &&
+  walletInstance.deviceDetails.platform === "android" &&
+  !!walletInstance.deviceDetails.x509Chain;
+
+const validateKeyAttestationChain = async (
+  walletInstance: WalletInstanceValid,
+  attestationServiceConfiguration: AttestationServiceConfiguration,
+) => {
+  // Check if device details exist and the platform is Android with a valid x509 chain. For iOS is not necessary.
+  if (hasAndroidCertificatesChain(walletInstance)) {
+    const x509Chain = walletInstance.deviceDetails.x509Chain.map(
+      (cert) => new X509Certificate(cert),
+    );
+
+    // Validate issuance and expiration of the certificates
+    validateIssuance(
+      x509Chain,
+      attestationServiceConfiguration.googlePublicKey,
+    );
+    // Validate revocation of the certificates using the provided CRL URL
+    await validateRevocation(
```

Comment:
this may throw, is that ok?

Reply:
yes the whole function can throw errors which are then caught by tryCatch